ftdc/ftdc.go

```diff
@@ -355,23 +369,85 @@ func (ftdc *FTDC) writeDatum(datum datum) error {
 // getWriter returns an io.Writer xor error for writing schema/data information. `getWriter` is only
 // expected to be called by `newDatum`.
 func (ftdc *FTDC) getWriter() (io.Writer, error) {
-	if ftdc.outputWriter != nil {
+	// If we have an `outputWriter` without a `currOutputFile`, it means ftdc was constructed with
+	// an explicit writer. We will use the passed in writer for all operations.
+	if ftdc.outputWriter != nil && ftdc.currOutputFile == nil {
 		return ftdc.outputWriter, nil
 	}
 
+	// Note to readers, until this function starts mutating `outputWriter` and `currOutputFile`, you
+	// can safely assume:
+	//
+	//   `outputWriter == nil if and only if currOutputFile == nil`.
+	//
+	// In case that helps reading the following logic.
+
+	// If we have an active outputWriter and we have not exceeded our FTDC file rotation quota, we
+	// can just return.
+	if ftdc.outputWriter != nil && ftdc.bytesWrittenCounter.count < ftdc.maxFileSize {
+		return ftdc.outputWriter, nil
+	}
+
+	// If we're in the logic branch where we have exceeded our FTDC file rotation quota, we first
+	// close the `currOutputFile`.
+	if ftdc.currOutputFile != nil {
+		ftdc.currOutputFile.Close()
+	}
+
 	var err error
-	ftdc.currOutputFile, err = os.Create("./viam-server.ftdc")
-	if err != nil {
+	// It's unclear in what circumstance we'd expect creating a new file to fail. Try 5 times for no
+	// good reason before giving up entirely and shutting down FTDC.
+	for numTries := 0; numTries < 5; numTries++ {
+		now := time.Now().UTC()
+		ftdc.currOutputFile, err = os.OpenFile(path.Join(ftdc.ftdcDir,
+			// Filename example: viam-server-2024-10-04T18-42-02.ftdc
+			fmt.Sprintf("viam-server-%d-%02d-%02dT%02d-%02d-%02dZ.ftdc",
+				now.Year(), now.Month(), now.Day(), now.Hour(), now.Minute(), now.Second())),
+			os.O_RDWR|os.O_CREATE|os.O_EXCL, 0644)
+		if err == nil {
+			break
+		}
 		ftdc.logger.Warnw("FTDC failed to open file", "err", err)
+
+		// If the error is some unexpected filename collision, wait a second to change the filename.
+		time.Sleep(time.Second)
+	}
+	if err != nil {
 		return nil, err
 	}
 
+	// New file, reset the bytes written counter.
+	ftdc.bytesWrittenCounter.count = 0
+
+	// When we create a new file, we must rewrite the schema. If we do not, a file may be useless
+	// without its "ancestors".
+	//
+	// As a hack, we decrement the `outputGenerationID` to force a new schema to be written.
```

Comment:
Ah this I do not fully understand. Could you explain why this hack works?

Reply:
Maybe we need to speak offline about this. I think the intuition of why this works is laid out in existing comments. I can certainly give a formal proof why this hack is a correct algorithm, but I don't think that's going to be useful.

I'm not surprised this isn't intuitive to you yet. I think I want to go down the route of talking synchronously about it. And then having you recommend what the documentation should be.

Reply:
My understanding is: decrementing `outputGenerationID` will mean `outputGenerationID != inputGenerationID` even if they were previously equal/there was no relevant schema change. That inequality will force FTDC to write a new schema to the top of this new file on the next call to `writeDatum`.

If that understanding is correct, then I think there's no need for more documentation here; if it's not then we may need to meet synchronously 😆 .

Reply:
Exactly!

And because outputGenerationId always trails (or equals) the inputGenerationId, subtracting by one never "accidentally" creates an equality. Whereas adding 1 to the `outputGenerationId` would be a bug waiting to happen.